lib/test-pools/single-process-test-pool.js

```diff
@@ -0,0 +1,175 @@
+			if (this.api.bailed) {
+				return;
+			}
+
+			const worker = this._worker(file, this.workerOptions);
+			this.runStatus.observeWorker(worker, file);
+
+			this.pendingWorkers.add(worker);
+			worker.promise.then(() => {
+				this.pendingWorkers.delete(worker);
+			});
+			this.restartTimer();
+
+			return worker.promise;
+		}, {concurrency: 1});
+	}
+
+	_worker(file, options) {
+		// Clear any global variables from previous test
+		revertGlobal();
+
+		const runner = new Runner({
+			file,
+			scheduledStart: true,
+			failFast: options.failFast,
+			failWithoutAssertions: options.failWithoutAssertions,
+			match: options.match,
+			projectDir: options.projectDir,
+			runOnlyExclusive: options.runOnlyExclusive,
+			serial: options.serial,
+			snapshotDir: options.snapshotDir,
+			updateSnapshots: options.updateSnapshots
+		});
+
+		let _reject;
+		let _resolve;
+
+		// Create promise to signify when test is finished
+		const promise = new Promise((resolve, reject) => {
+			_resolve = resolve;
+			_reject = reject;
+		});
+
+		// Map exit to resolve or reject
+		const exit = code => {
+			process.removeListener('unhandledRejection', unhandledRejectionHandler);
+			runner.interrupt();
+			if (code && options.failFast) {
+				_reject();
+			} else {
+				_resolve();
+			}
+		};
+
+		let statusListener;
+		const onStateChange = listener => {
+			statusListener = evt => {
+				evt.testFile = file;
+				listener(evt);
+			};
+
+			return runner.on('stateChange', statusListener);
+		};
+
+		const notifyOfPeerFailure = () => {
+			runner.interrupt();
+		};
+
+		const attributedRejections = new Set();
+		const unhandledRejectionHandler = (reason, promise) => {
+			if (runner.attributeLeakedError(reason)) {
+				attributedRejections.add(promise);
+			}
+		};
+
+		process.on('unhandledRejection', unhandledRejectionHandler);
+
+		runner.on('error', error => {
+			statusListener({type: 'internal-error', err: serializeError('Internal runner error', false, error)});
+			exit(1);
+		});
+
+		runner.on('finish', () => {
+			try {
+				const touchedFiles = runner.saveSnapshotState();
+				if (touchedFiles) {
+					statusListener({type: 'touched-files', files: touchedFiles});
+				}
+			} catch (error) {
+				statusListener({type: 'internal-error', err: serializeError('Internal runner error', false, error)});
+				exit(1);
+				return;
+			}
+
+			nowAndTimers.setImmediate(() => {
+				currentlyUnhandled()
+					.filter(rejection => !attributedRejections.has(rejection.promise))
+					.forEach(rejection => {
+						statusListener({type: 'unhandled-rejection', err: serializeError('Unhandled rejection', true, rejection.reason)});
+					});
+
+				exit(0);
+			});
+		});
+
+		let accessedRunner = false;
+		global.getRunner = () => {
+			accessedRunner = true;
+			return runner;
+		};
+
+		// Store value in case to prevent required modules from modifying it.
+		const testPath = file;
+
+		// Give runStatus a chance to setup listener
+		setTimeout(() => {
```

Comment:
I believe this should use `nowAndTimers.setTimeout`, otherwise use of `sinon.useFakeTimers` and forgetting to restore the original timer functions in a test can cause the test run to either hang inexplicably or exit without completing (a difficult to debug problem as I discovered!)

Reply:
Good catch, I actually updated it to use `nowAndTimers.setImmediate` since it just needs to run after `runStatus.observeWorker` has been called.